### ISSUE TITLE
Stack overflow inside rustc_middle::ty::print::Printer

### ISSUE THREAD

**Issue by Asapin** [rust-lang/rust#88664]
Getting a stack overflow error even on a simple `cargo check` if there's too many structs inside the implementation of `serde::Deserialize` trait.

```
cargo check
    Checking err v0.1.0 (D:\projects\rust\err)

thread 'rustc' has overflowed its stack
error: could not compile `err`                                                                                                                               

Caused by:
  process didn't exit successfully: `rustc --crate-name err --edition=2018 src\main.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,metadata -C embed-bitcode=no -C debuginfo=2 -C metadata=9fd35313c6199767 -C extra-filename=-9fd35313c6199767 --out-dir D:\projects\rust\err\target\debug\deps -C incremental=D:\projects\rust\err\target\debug\incremental -L dependency=D:\projects\rust\err\target\debug\deps --extern serde=D:\projects\rust\err\target\debug\deps\libserde-5ecaf821eaf01989.rmeta --extern serde_json=D:\projects\rust\err\target\debug\deps\libserde_json-ec8cae54bf55cb2a.rmeta` (exit code: 0xc00000fd, STATUS_STACK_OVERFLOW)
```

Project to reproduce: [link to playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9d69c882791d8d8a1b26d31e8be784c2)

But if I move all temporary structs out of the `fn deserialize<D>` method, the error goes away: [working version](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7680bc6f791d82d61364657cbe5caa6b)

Crash dump from WinDbg: [~240MB](https://drive.google.com/file/d/1pl9ImbWXJhDwUffWe4t0z1yXSPpCEmur/view?usp=sharing)

As far as I can tell, stack overflow happens somewhere inside `rustc_middle::ty::print::Printer`

**Comment by hellow554** [comment]
Looks like a dup of https://github.com/rust-lang/rust/issues/87932, but I'm not sure yet

**Comment by Asapin** [comment]
Yes, it does look very similar. Which is weird as I have almost identical to #87932 code in another place except the amount of enums and structs inside the `fn deserialize` (2 enums and 4 structs to be precise), and it doesn't cause any errors.

**Comment by tmiasko** [comment]
Fixed in #89738.
